### PR TITLE
KAZOO-5449: reset api for private key, identity secrets and get public key

### PR DESCRIPTION
**Pull request by icehess**
and KAZOO-5450

### CODE REVIEW

File: applications/crossbar/src/modules/cb_auth.erl [hunk 1]
```diff
@@ -189,53 +225,57 @@ validate_path(Context, ?TOKENINFO_PATH, ?HTTP_POST) ->
...
-validate_path(Context, ?LINKS_PATH, Id, ?HTTP_GET) ->
+validate_path_id(Context, ?LINKS_PATH, Id, ?HTTP_GET) ->
+    crossbar_doc:load(Id, Context, ?TYPE_CHECK_OPTION(<<"user">>));
+validate_path_id(Context, ?LINKS_PATH, Id, ?HTTP_PUT) ->
     crossbar_doc:load(Id, Context, ?TYPE_CHECK_OPTION(<<"user">>));
-validate_path(Context, ?LINKS_PATH, Id, ?HTTP_PUT) ->
+validate_path_id(Context, ?LINKS_PATH, Id, ?HTTP_DELETE) ->
     crossbar_doc:load(Id, Context, ?TYPE_CHECK_OPTION(<<"user">>));
-validate_path(Context, ?LINKS_PATH, Id, ?HTTP_DELETE) ->
-    crossbar_doc:load(Id, Context, ?TYPE_CHECK_OPTION(<<"user">>)).
```
Comment (lazedo) [inline applications/crossbar/src/modules/cb_auth.erl]: why the changes above ?

File: applications/crossbar/src/modules/cb_auth.erl [hunk 2]
```diff
@@ -189,53 +225,57 @@ validate_path(Context, ?TOKENINFO_PATH, ?HTTP_POST) ->
...
+validate_path_id(Context, ?KEYS_PATH, ?PRIVATE_PATH, ?HTTP_PATCH) ->
+    reset_system_private_key(Context);
+validate_path_id(Context, ?KEYS_PATH, ?PUBLIC_PATH, ?HTTP_GET) ->
+    get_system_public_key(Context).
```
Comment (lazedo) [inline applications/crossbar/src/modules/cb_auth.erl]: as commented above, use keys/:keyid GET/DELETE

File: applications/crossbar/src/modules/cb_auth.erl [hunk 3]
```diff
@@ -372,3 +412,121 @@ add_app(Context) ->
...
+    case kz_auth_identity:reset_system_secret() of
+        {'ok', _} -> cb_context:set_resp_status(Context, 'success');
+        {'error', _Reason} ->
+            lager:warning("failed to reset system identity secret: ~p", [_Reason]),
+            cb_context:add_system_error('datastore_fault', Context)
+    end.
+
+-spec reset_identity_secret(cb_context:context()) -> cb_context:context().
+reset_identity_secret(Context) ->
+    OwnerId = kz_json:get_ne_binary_value(<<"owner_id">>, cb_context:doc(Context)),
```
Comment (lazedo) [inline applications/crossbar/src/modules/cb_auth.erl]: use cb_context:owner_id , getting owner_id from the doc does not ensure the path

File: core/kazoo_auth/src/kz_auth_jwt.erl [hunk 4]
```diff
@@ -41,11 +41,11 @@ decode(JWTToken, Verify) when is_binary(JWTToken) ->
     end;
 decode(#{header := Header, payload := Payload} = Token, 'true') ->
     case verify(Token) of
-        'true' -> {'ok', kz_json:from_map(Header), kz_json:from_map(Payload)};
+        'true' -> {'ok', maps:to_list(Header), maps:to_list(Payload)};
```
Comment (lazedo) [inline core/kazoo_auth/src/kz_auth_jwt.erl]: this needs strong verification that it doesn't break sso / oauth integration

File: core/kazoo_auth/src/kz_auth_jwt.erl [hunk 5]
```diff
@@ -41,11 +41,11 @@ decode(JWTToken, Verify) when is_binary(JWTToken) ->
...
         'false' -> {'error', 'verify_failed'}
     end;
 decode(#{header := Header, payload := Payload}, 'false') ->
-    {'ok', kz_json:from_map(Header), kz_json:from_map(Payload)}.
+    {'ok', maps:to_list(Header), maps:to_list(Payload)}.
```
Comment (lazedo) [inline core/kazoo_auth/src/kz_auth_jwt.erl]: this needs strong verification that it doesn't break sso / oauth integration

File: core/kazoo_auth/src/kz_auth_keys.erl [hunk 6]
```diff
@@ -281,8 +286,9 @@ save_private_key(JObj, Key) ->
               ,{'content_type', ?SYSTEM_KEY_ATTACHMENT_CTYPE}
               ],
     case kz_datamgr:put_attachment(?KZ_AUTH_DB, KeyId, ?SYSTEM_KEY_ATTACHMENT_NAME, to_pem(Key), Options) of
-        {'ok', _} -> store({'private', KeyId}, Key),
-                     {'ok', Key};
+        {'ok', _} ->
+            store({'private', KeyId}, Key),
+            {'ok', Key};
```
Comment (lazedo) [inline core/kazoo_auth/src/kz_auth_keys.erl]: what changed here ?

File: core/kazoo_auth/src/kz_auth_keys.erl [hunk 7]
```diff
@@ -308,3 +314,60 @@ gen_private_key() ->
...
+        {'ok', _} ->
+            lager:debug("refreshing private/public cache for key ~s", [KeyId]),
+
+            store({'private', KeyId}, PrivateKey),
+            PublicKey = get_public_key_from_private_key(PrivateKey),
+            store(KeyId, PublicKey);
+        {'error', _Err}=Err ->
+            lager:debug("error ~p saving generated system key ~s", [_Err, KeyId]),
+            Err
+    end.
```
Comment (lazedo) [inline core/kazoo_auth/src/kz_auth_keys.erl]: the above code is duplication of existing code. the `reset` can be achieved by setting `?SYSTEM_KEY_ID` to a new rand_hex

File: applications/crossbar/src/modules/cb_auth.erl [hunk 8]
```diff
@@ -113,20 +138,33 @@ authorize(Context, PathToken) ->
...
-    cb_context:is_authenticated(Context);
-authorize_nouns(Context, ?LINKS_PATH, _Id, ?HTTP_DELETE, [{<<"auth">>, _}]) ->
-    cb_context:is_authenticated(Context).
+-spec authorize_nouns_id(cb_context:context(), path_token(), path_token(), req_verb(), req_nouns()) -> boolean().
+authorize_nouns_id(_, ?APPS_PATH,      _Id,           ?HTTP_GET,    [{<<"auth">>, _}]) -> 'true';
+authorize_nouns_id(C, ?KEYS_PATH,      ?PRIVATE_PATH, ?HTTP_PATCH,  [{<<"auth">>, _}]) -> cb_context:is_superduper_admin(C);
+authorize_nouns_id(C, ?KEYS_PATH,      ?PUBLIC_PATH,  ?HTTP_GET,    [{<<"auth">>, _}]) -> cb_context:is_account_admin(C);
+authorize_nouns_id(_, ?LINKS_PATH,     _Id,           ?HTTP_GET,    [{<<"auth">>, _}]) -> 'true';
+%% monster-ui still uses this (accounts/123/auth/links)
+authorize_nouns_id(_, ?LINKS_PATH,     _Id,           ?HTTP_GET,    [{<<"auth">>, _}, [<<"accounts">>, _]]) -> 'true';
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_auth.erl]: I think it should be `[{auth, _}, {accounts, _}]`, you have square brackets around accounts

File: applications/crossbar/src/modules/cb_auth.erl [hunk 9]
```diff
@@ -88,45 +96,78 @@ allowed_methods(?KEYS_PATH, _KeyId) -> [?HTTP_GET , ?HTTP_POST , ?HTTP_DELETE].
...
-authorize_nouns(Context, ?LINKS_PATH, _Id, ?HTTP_PUT, [{<<"auth">>, _}]) ->
-    cb_context:is_authenticated(Context);
-authorize_nouns(Context, ?LINKS_PATH, _Id, ?HTTP_DELETE, [{<<"auth">>, _}]) ->
-    cb_context:is_authenticated(Context).
+authorize_nouns(_, ?APPS_PATH,      _Id,           ?HTTP_GET,    [{<<"auth">>, _}]) -> 'true';
+authorize_nouns(C, ?KEYS_PATH,      _Id,           ?HTTP_GET,    [{<<"auth">>, _}]) -> cb_context:is_account_admin(C);
+authorize_nouns(C, ?KEYS_PATH,      _Id,           ?HTTP_PATCH,  [{<<"auth">>, _}]) -> cb_context:is_superduper_admin(C);
+authorize_nouns(_, ?LINKS_PATH,     _Id,           ?HTTP_GET,    [{<<"auth">>, _}]) -> 'true';
+%% monster-ui still uses this (accounts/123/auth/links)
+authorize_nouns(_, ?LINKS_PATH,     _Id,           ?HTTP_GET,    [{<<"auth">>, _}, [<<"accounts">>, _]]) -> 'true';
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_auth.erl]: This is wrong again. Use curly instead of square brackets
